File: crap4py/scripts/crapkit/score.py
from dataclasses import dataclass

CRAPPY_THRESHOLD = 30.0


def crap_score(complexity: int, coverage: float | None) -> float | None:
    if coverage is None:
        return None
    uncovered = 1.0 - coverage / 100.0
    return complexity**2 * uncovered**3 + complexity


@dataclass(frozen=True)
class Entry:
    name: str
    module: str
    file: str
    start_line: int
    end_line: int
    complexity: int
    coverage: float | None
    statements: int
    covered_statements: int

    @property
    def crap(self) -> float | None:
        return crap_score(self.complexity, self.coverage)


def sort_entries(entries: list[Entry]) -> list[Entry]:
    def key(entry: Entry) -> tuple[bool, float, str]:
        score = entry.crap
        return (score is None, -score if score is not None else 0.0, entry.name)

    return sorted(entries, key=key)

File: crap4ts/scripts/lib/score.mjs
export const CRAPPY_THRESHOLD = 30;

export function crapScore(complexity, coverage) {
  if (coverage === null || coverage === undefined) {
    return null;
  }
  const uncovered = 1 - coverage / 100;
  return complexity ** 2 * uncovered ** 3 + complexity;
}

export function sortEntries(entries) {
  return [...entries].sort((left, right) => {
    const leftScore = crapScore(left.complexity, left.coverage);
    const rightScore = crapScore(right.complexity, right.coverage);
    if (leftScore === null && rightScore === null) {
      return left.name.localeCompare(right.name);
    }
    if (leftScore === null) return 1;
    if (rightScore === null) return -1;
    if (leftScore !== rightScore) return rightScore - leftScore;
    return left.name.localeCompare(right.name);
  });
}
